src/pudl/extract/excel.py

```diff
@@ -0,0 +1,254 @@
+        if bad_years:
+            raise ValueError(
+                f"Requested invalid years for {self._dataset_name}: {bad_years}\n"
+                f"Supported years: {pc.working_years[self._dataset_name]}\n"
+            )
+
+        excel_files = {}
+        for file_path in self._get_all_file_paths(years):
+            logger.info(
+                f'Loading excel spreadsheet from {file_path}')
+            excel_files[file_path] = pd.ExcelFile(file_path)
+
+        # excel_files now contains pre-loaded excel files, now munch the data
+        # per page and put them in raw_dfs[page] = DataFrame
+        raw_dfs = {}
+        for page in self._metadata.get_all_pages():
+            if page in self.BLACKLISTED_PAGES:
+                logger.info(f'Skipping blacklisted page {page}.')
+                continue
+            df = pd.DataFrame()
+            for yr in years:
+                data = excel_files[self._get_file_path(yr, page)]
+
+                logger.info(
+                    f'Loading dataframe for {self._dataset_name} {page} {yr}')
+                newdata = pd.read_excel(
+                    data,
+                    sheet_name=self._metadata.get_sheet_name(yr, page),
+                    skiprows=self._metadata.get_skiprows(yr, page),
+                    dtype=self.get_dtypes(yr, page))
+
+                newdata = pudl.helpers.simplify_columns(newdata)
+                newdata = self.process_raw(yr, page, newdata)
+                newdata = newdata.rename(
+                    columns=self._metadata.get_column_map(yr, page))
+                newdata = self.process_renamed(yr, page, newdata)
+                df = df.append(newdata, sort=True)
+
+            # After all years are loaded, consolidate missing columns
+            missing_cols = self._metadata.get_all_columns(
+                page).difference(df.columns)
+            empty_cols = pd.DataFrame(columns=missing_cols)
+            df = pd.concat([df, empty_cols], sort=True)
+            raw_dfs[page] = self.process_final_page(page, df)
+        return raw_dfs
+
+    def _get_all_file_paths(self, years):
+        """Returns list of all data files that will be loaded for all years."""
+        bad_years = []
+        all_files = []
+        for page in self._metadata.get_all_pages():
+            for yr in years:
+                try:
+                    all_files.append(self._get_file_path(yr, page))
+                except IndexError:
+                    bad_years.append(yr)
+        if bad_years:
+            raise FileNotFoundError(
+                f'Missing {self._dataset_name} files for years {bad_years}.')
+        return all_files
+
+    def verify_years(self, years):
+        """Validate that all files are availabe.
+
+        Raises:
+            FileNotFoundError: when some files are not found in the datastore.
+        """
+        self._get_all_file_paths(years)
+
+    def file_basename_glob(self, year, page):
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Method could be a function](https://app.codacy.com/manual/zaneselvans/pudl/pullRequest?prid=5191218)